frame/lending/src/tests_offchain.rs

```diff
@@ -0,0 +1,159 @@
+		// Create a market with BTC as collateral asset and USDT as borrow asset.
+		// Initial collateral asset price is 50_000 USDT. Market's collateral factor equals two.
+		// It means that borrow supposed to be undercolateraized when
+		// borrowed amount is higher then one half of collateral amount in terms of USDT.
+		let (market_id, vault) = create_market::<50_000>(
+			USDT::instance(),
+			BTC::instance(),
+			manager,
+			DEFAULT_MARKET_VAULT_RESERVE,
+			MoreThanOneFixedU128::saturating_from_integer::<u128>(2),
+		);
+
+		// Deposit collateral.
+		let collateral_value = BTC::units(1);
+		assert_ok!(Tokens::mint_into(BTC::ID, &manager, collateral_value));
+		assert_extrinsic_event::<Runtime>(
+			Lending::deposit_collateral(Origin::signed(manager), market_id, collateral_value),
+			Event::Lending(crate::Event::CollateralDeposited {
+				sender: manager,
+				amount: collateral_value,
+				market_id,
+			}),
+		);
+		// Deposit USDT in the vault.
+		let vault_value = USDT::units(100_000_000);
+		assert_ok!(Tokens::mint_into(USDT::ID, &lender, vault_value));
+		assert_ok!(Vault::deposit(Origin::signed(lender), vault, vault_value));
+		process_and_progress_blocks(1);
+		// Borrow 20_000 USDT.
+		let borrowed_value = USDT::units(20_000);
+		assert_extrinsic_event::<Runtime>(
+			Lending::borrow(Origin::signed(manager), market_id, borrowed_value),
+			Event::Lending(crate::Event::Borrowed {
+				sender: manager,
+				amount: borrowed_value,
+				market_id,
+			}),
+		);
+
+		// Emulate situation when collateral price has fallen down
+		// from 50_000 USDT to 38_000 USDT.
+		// Now the borrow is undercolateraized since market's collateral factor equals two.
+		// Therefore, one BTC can cover only 19_000 of 20_0000 borrowed USDT.
+		set_price(BTC::ID, NORMALIZED::units(38_000));
+		// Header for the fake block to execute off-chain worker
+		let header =
+			Header::new(2, H256::default(), H256::default(), [69u8; 32].into(), Digest::default());
+		// Execute off-chain worker
+		Executive::offchain_worker(&header);
+		// Check if corresponded liquidation transaction has been placed in the pool
+		let tx = pool_state.write().transactions.pop().unwrap();
+		assert!(pool_state.read().transactions.is_empty());
+		let tx = Extrinsic::decode(&mut &*tx).unwrap();
+		assert_eq!(
+			tx.call,
+			Call::Lending(crate::Call::liquidate { market_id, borrowers: vec![manager.clone()] })
+		);
+		process_block_with_execution(tx);
+		// Check event from Lending pallet
+		let event =
+			crate::Event::LiquidationInitiated { market_id, borrowers: vec![manager.clone()] };
+		System::assert_has_event(Event::Lending(event));
+		// Check event from Liquidations pallet
+		let event = pallet_liquidations::Event::PositionWasSentToLiquidation {};
+		System::assert_has_event(Event::Liquidations(event));
+	});
+}
+
+// HELPERS
+fn default_under_collateralized_warn_percent() -> Percent {
+	Percent::from_float(0.10)
+}
+
+fn create_market<const NORMALIZED_PRICE: u128>(
```

Comment:
it feels like `create_market<const NORMALIZED_PRICE> already exists in lending. why copy paste? can we parametrize types to avoid copy paste?

Reply:
Done